php-frontend/src/main/java/org/sonar/php/cache/Cache.java

```diff
@@ -19,37 +19,99 @@
     this.cacheContext = cacheContext;
+    this.pluginVersion = cacheContext.pluginVersion();
+    this.projectKey = cacheContext.projectKey();
   }
 
-  public void write(ProjectSymbolData projectSymbolData) {
+  public void write(InputFile file, SymbolTableImpl symbolTable) {
     if (cacheContext.isCacheEnabled()) {
-      String pluginVersion = cacheContext.pluginVersion();
-      String projectKey = cacheContext.projectKey();
-      SerializationInput serializationInput = new SerializationInput(projectSymbolData, pluginVersion);
-      SerializationResult serializationData = ProjectSymbolDataSerializer.toBinary(serializationInput);
-      cacheContext.getWriteCache().writeBytes(CACHE_KEY_DATA + projectKey, serializationData.data());
-      cacheContext.getWriteCache().writeBytes(CACHE_KEY_STRING_TABLE + projectKey, serializationData.stringTable());
+      SerializationInput serializationInput = new SerializationInput(symbolTable, pluginVersion);
+      SerializationResult serializationData = SymbolTableSerializer.toBinary(serializationInput);
+
+      String cacheFileName = getCacheFileName(file);
+      if (cachedFileNames.add(cacheFileName)) {
+        PhpWriteCache writeCache = cacheContext.getWriteCache();
+        writeCache.writeBytes(getDataCacheKey(cacheFileName), serializationData.data());
+        writeCache.writeBytes(getStringTableCacheKey(cacheFileName), serializationData.stringTable());
+      }
     }
   }
 
-  public ProjectSymbolData read() {
+  @CheckForNull
+  public SymbolTableImpl read(InputFile file) {
     if (cacheContext.isCacheEnabled()) {
-      String pluginVersion = cacheContext.pluginVersion();
-      String projectKey = cacheContext.projectKey();
-      byte[] data = cacheContext.getReadCache().readBytes(CACHE_KEY_DATA + projectKey);
-      byte[] stringTable = cacheContext.getReadCache().readBytes(CACHE_KEY_STRING_TABLE + projectKey);
-      return ProjectSymbolDataDeserializer.fromBinary(new DeserializationInput(data, stringTable, pluginVersion));
+      String cacheFileName = getCacheFileName(file);
+      byte[] data = cacheContext.getReadCache().readBytes(getDataCacheKey(cacheFileName));
+      byte[] stringTable = cacheContext.getReadCache().readBytes(getStringTableCacheKey(cacheFileName));
+      if (data != null && stringTable != null) {
+        return SymbolTableDeserializer.fromBinary(new DeserializationInput(data, stringTable, pluginVersion));
+      }
     }
     return null;
   }
+
+  private String getDataCacheKey(String cacheFileName) {
+    return CACHE_KEY_DATA + projectKey + ":" + cacheFileName;
+  }
+
+  private String getStringTableCacheKey(String cacheFileName) {
+    return CACHE_KEY_STRING_TABLE + projectKey + ":" + cacheFileName;
+  }
+
+  public String getCacheFileName(InputFile file) {
+    MessageDigest digest = getMessageDigest();
+    digest.update(computeFileHash(file).toByteArray());
+    return new BigInteger(1, digest.digest()).toString();
```

Comment:
Why not simple `return new String(digest.digest());` ?

Reply:
It produces a String looking like `0���DZ�8RP��0�N�v�P!5RP�Nk` instead of `22146621043327621015461537661717798110946814184349681185080520559963237469803`